bundles/org.openhab.binding.monopriceaudio/src/main/java/org/openhab/binding/monopriceaudio/internal/handler/MonopriceAudioHandler.java

```diff
@@ -0,0 +1,706 @@
+                new ChannelUID(getThing().getUID(), ALL + CHANNEL_DELIMIT + CHANNEL_TYPE_ALLSOURCE), sourcesLabels);
+        activeZones.forEach(zoneNum -> {
+            stateDescriptionProvider.setStateOptions(new ChannelUID(getThing().getUID(),
+                    ZONE.toLowerCase() + zoneNum + CHANNEL_DELIMIT + CHANNEL_TYPE_SOURCE), sourcesLabels);
+        });
+
+        // remove the channels for the zones we are not using
+        if (numZones < MAX_ZONES) {
+            List<Channel> channels = new ArrayList<>(this.getThing().getChannels());
+
+            List<Integer> zonesToRemove = IntStream.range(numZones + 1, MAX_ZONES + 1).boxed()
+                    .collect(Collectors.toList());
+
+            zonesToRemove.forEach(zone -> {
+                channels.removeIf(c -> (c.getUID().getId().contains(ZONE.toLowerCase() + zone)));
+            });
+            updateThing(editThing().withChannels(channels).build());
+        }
+
+        // initialize the all volume state
+        allVolume = initialAllVolume;
+        long allVolumePct = Math
+                .round((double) (initialAllVolume - MIN_VOLUME) / (double) (MAX_VOLUME - MIN_VOLUME) * 100.0);
+        updateState(ALL + CHANNEL_DELIMIT + CHANNEL_TYPE_ALLVOLUME, new PercentType(BigDecimal.valueOf(allVolumePct)));
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        scheduleReconnectJob();
+        schedulePollingJob();
+    }
+
+    @Override
+    public void dispose() {
+        cancelReconnectJob();
+        cancelPollingJob();
+        closeConnection();
+        ignoreZones.clear();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        String channel = channelUID.getId();
+        String[] channelSplit = channel.split(CHANNEL_DELIMIT);
+        MonopriceAudioZone zone = MonopriceAudioZone.valueOf(channelSplit[0].toUpperCase());
+        String channelType = channelSplit[1];
+
+        if (getThing().getStatus() != ThingStatus.ONLINE) {
+            logger.debug("Thing is not ONLINE; command {} from channel {} is ignored", command, channel);
+            return;
+        }
+
+        boolean success = true;
+        synchronized (sequenceLock) {
+            if (!connector.isConnected()) {
+                logger.debug("Command {} from channel {} is ignored: connection not established", command, channel);
+                return;
+            }
+
+            Stream<String> zoneStream = MonopriceAudioZone.VALID_ZONES.stream().limit(numZones);
+            try {
+                switch (channelType) {
+                    case CHANNEL_TYPE_POWER:
+                        if (command instanceof OnOffType) {
+                            connector.sendCommand(zone, MonopriceAudioCommand.POWER, command == OnOffType.ON ? 1 : 0);
+                            ZONE_DATA_MAP.get(zone.getZoneId()).setPower(command == OnOffType.ON ? ON_STR : OFF_STR);
+                        }
+                        break;
```

Comment:
You should try to handle the RefreshType command for all of these channels.

Reply:
If you can't query the channel data, then you should at least used cached values.

Reply:
Done and pushed.